pkg/sfu/receiver.go

```diff
@@ -209,6 +216,21 @@ func (w *WebRTCReceiver) AddUpTrack(track *webrtc.TrackRemote, buff *buffer.Buff
 		tracker.OnStatusChanged(func(status StreamStatus) {
 			if status == StreamStatusStopped {
 				w.removeAvailableLayer(uint16(layer))
+
+				//
+				// If the layer is expected to stop, reset tracker so that it starts up
+				// on the first packet of the layer on restart.
+				//
+				// As there may be packets from client before a layer expected to stop actually stops,
+				// use a window in which a status stopped is treated as an expected stop.
+				// If the stop happens outside the window, treat it like an unexpected stop.
+				//
+				maxExpectedLayer, ok := w.maxExpectedLayer.Load().(MaxExpectedSpatialLayer)
+				if ok {
+					if time.Since(maxExpectedLayer.setAt) < 30*time.Second && layer > maxExpectedLayer.layer {
```

Comment:
instead of the time window, would it be better if we called a function on receiver right before resuming a layer on the server side? it seems more predictable and we could get rid of the timing here.

Reply:
Will look at moving this on expected start.

Reply:
Would this be some thing like the following?
- When max subscribed layer is decreasing, we do not have to do anything. If the client implements stopping layers, the layer will stop eventually and stream tracker will catch it and the rest flows like now.
- When max subscribed layer is increasing, two things need to be handled
   - If clients did not stop any layer, we do not need to reset the tracker.
   - If the clients did stop a layer, we need to reset.
   So, maybe, we cross check with currently available layers and reset layers which are not available now, but is expected to start shortly? That should cover both cases.
   
Does that make sense?

Reply:
ah yes, that's a good case you are mentioning. we'd need to check the current layer state to see if a reset is necessary. 👍

Reply:
@davidzhao Have addressed your comments in this commit - https://github.com/livekit/livekit-server/pull/295/commits/973e2b400381018afe64501b34f3334da67093c1. Also set the first update to be after 10 seconds. Seemed like a good enough wait to do that.

Reply:
Added this commit as well for proper locking - https://github.com/livekit/livekit-server/pull/295/commits/1a79c18e90354151965c7029d4a4a8ee453d7347

Reply:
One more :-) A test was crashing because of nil pointer dereference. The channel for the `updateStats` timer gets closed after `t.receiver` is set to `nil` - https://github.com/livekit/livekit-server/pull/295/commits/3284225d283b635c5273240b5dad6d561ce7c71a